drivers/adc/adc_npcx.c

```diff
@@ -0,0 +1,374 @@
+
+	adc_context_lock(&data->ctx, false, NULL);
+	error = adc_npcx_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+#if defined(CONFIG_ADC_ASYNC)
+static int adc_npcx_read_async(const struct device *dev,
+			      const struct adc_sequence *sequence,
+			      struct k_poll_signal *async)
+{
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	int error;
+
+	adc_context_lock(&data->ctx, true, async);
+	error = adc_npcx_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+/* ADC driver registration */
+static const struct adc_driver_api adc_npcx_driver_api = {
+	.channel_setup = adc_npcx_channel_setup,
+	.read = adc_npcx_read,
+#if defined(CONFIG_ADC_ASYNC)
+	.read_async = adc_npcx_read_async,
+#endif
+	.ref_internal = NPCX_ADC_VREF_VOL,
+};
+
+static int adc_npcx_init(const struct device *dev);
+
+static const struct npcx_alt adc_alts[] = DT_NPCX_ALT_ITEMS_LIST(0);
+
+static const struct adc_npcx_config adc_npcx_cfg_0 = {
+	.base = DT_INST_REG_ADDR(0),
+	.clk_cfg = DT_NPCX_CLK_CFG_ITEM(0),
+	.alts_list = adc_alts,
+};
+
+static struct adc_npcx_data adc_npcx_data_0 = {
+	ADC_CONTEXT_INIT_TIMER(adc_npcx_data_0, ctx),
+	ADC_CONTEXT_INIT_LOCK(adc_npcx_data_0, ctx),
+	ADC_CONTEXT_INIT_SYNC(adc_npcx_data_0, ctx),
+};
+
+DEVICE_AND_API_INIT(adc_npcx, DT_INST_LABEL(0),
+		    adc_npcx_init,
+		    &adc_npcx_data_0, &adc_npcx_cfg_0,
+		    PRE_KERNEL_1,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		    &adc_npcx_driver_api);
+
+static int adc_npcx_init(const struct device *dev)
+{
+	const struct adc_npcx_config *const config = DRV_CONFIG(dev);
+	struct adc_npcx_data *const data = DRV_DATA(dev);
+	struct adc_reg *const inst = HAL_INSTANCE(dev);
+	const struct device *const clk_dev =
+					device_get_binding(NPCX_CLK_CTRL_NAME);
+	int prescaler = 0;
+
+	/* Save ADC device in data */
+	data->adc_dev = dev;
+
+	/* Turn on device clock first and get source clock freq. */
+	if (clock_control_on(clk_dev,
+		(clock_control_subsys_t *)&config->clk_cfg) != 0) {
+		LOG_ERR("Turn on ADC clock fail.");
+		return -EIO;
+	}
+
+	if (clock_control_get_rate(clk_dev, (clock_control_subsys_t *)
+			&config->clk_cfg, &data->input_clk) != 0) {
+		LOG_ERR("Get ADC clock rate error.");
+		return -EIO;
+	}
+
+	/* Configure the ADC clock */
+	prescaler = ceiling_fraction(data->input_clk, NPCX_ADC_CLK);
+	if (prescaler > 0x40)
+		prescaler = 0x40;
+
+	/* Set Core Clock Division Factor in order to obtain the ADC clock */
+	SET_FIELD(inst->ATCTL, NPCX_ATCTL_SCLKDIV_FIELD, prescaler - 1);
+
+	/* Set regular ADC delay */
+	SET_FIELD(inst->ATCTL, NPCX_ATCTL_DLY_FIELD, ADC_REGULAR_DLY_VAL);
+
+	/* Set ADC speed sequentially */
+	inst->ADCCNF2 = ADC_REGULAR_ADCCNF2_VAL;
+	inst->GENDLY = ADC_REGULAR_GENDLY_VAL;
+	inst->MEAST = ADC_REGULAR_MEAST_VAL;
+
+	/* Configure ADC interrupt and enable it */
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority),
+			    adc_npcx_isr,
```

Comment:
It's hard for me to tell, but this line should line up with the 'D' after IRQ_CONNECT(

Reply:
Done.